### DISCUSSION TITLE
MPL license "Incompatible With Secondary Licenses"

### DISCUSSION THREAD

**Discussion by link2xt** [uxbox/uxbox#813] [Q&A] [answered]
I searched the issues and discussions but didn't find any mention of the license.

Why choose MPLv2 variant "Incompatible With Secondary Licenses" rather than compatible with secondary licenses (if you want weak copyleft) or GPL/AGPL (if you want copyleft)? As far as I know the "Incompatible With Secondary Licenses" variant is mainly used for compatibility with MPLv1.1, when MPLv1.1 project that was never dual-licensed with LGPL transitions to MPLv2. But why choose incompatible MPLv2 license for new software? It's basically GPL-like copyleft but can't share the code with GPL software.

**Comment by niwinz** [comment, marked as answer]
Hi @link2xt 

The main reason we made this choide is because we don't want a **Larger Work** distributed using exclusivelly with GPL.

> 3.3. Distribution of a Larger Work
> You may create and distribute a Larger Work under terms of Your choice, provided that You also comply with the requirements of this License for the Covered Software. If the Larger Work is a combination of Covered Software with a work governed by one or more Secondary Licenses, and the Covered Software is not Incompatible With Secondary Licenses, this License permits You to additionally distribute such Covered Software under the terms of such Secondary License(s), **so that the recipient of the Larger Work may, at their option, further distribute the Covered Software under the terms of either this License or such Secondary License(s)**.

Maybe I'm wrong but as far as I understand it, without using the **Exhibit B**, changes made to a larger work and then distributed with GPL (because the recipient can choice distribute ALL under GPL) can't be ported back to penpot without infecting it with GPL.

**Reply by link2xt** [reply]
> Maybe I'm wrong but as far as I understand it, without using the Exhibit B, changes made to a larger work and then distributed with GPL (because the recipient can choice distribute ALL under GPL) can't be ported back to penpot without infecting it with GPL.

This is not true according to the answer to Q14 in MPLv2 FAQ: https://www.mozilla.org/en-US/MPL/2.0/FAQ/
It says that if someone builds a new GPLed program based on your MPLv2 code (this happens when you compile a binary out of added GPL code and existing MPLv2 files), they must still make modifications of MPLv2 code available separately under MPLv2 license. They will have to keep these files with MPLv2 header in their codebase, not simply relicense it under GPL without contributing back to MPLv2 codebase.

The only thing this clause prevents is compiling a GPL binary. But your version of MPLv2 still allows someone to do a worse thing: use your code in a proprietary binary while only releasing the source code of modified files. Additional proprietary code may reside in other files which are not required to be released in source code form. If you want to make sure you can get the changes back to this codebase, why not release under GPL or AGPL in the first place?

Even if you are ok with someone building larger proprietary binary based on your source, using this variant of MPLv2 does not make sense as it only restricts someone from building GPL binaries, but not proprietary binaries.

As far as I understand this variant of MPLv2 (with Exhibit B) was not intended to be used in the new code. The answer to "Q4: I want to use the Mozilla Public License for software that I have written. What do I have to do?" of the FAQ simply says to use Exhibit A. Exhibit B only exists to make it possible to update legacy MPLv1.1-only code which was not dual-licensed under MPLv1.1/LGPL to MPLv2 and thus avoid the complexity of dealing with legacy MPLv1.1 forever.

**Reply by niwinz** [reply]
> This is not true according to the answer to Q14 in MPLv2 FAQ: https://www.mozilla.org/en-US/MPL/2.0/FAQ/
It says that if someone builds a new GPLed program based on your MPLv2 code (this happens when you compile a binary out of added GPL code and existing MPLv2 files), they must still make modifications of MPLv2 code available separately under MPLv2 license. They will have to keep these files with MPLv2 header in their codebase, not simply relicense it under GPL without contributing back to MPLv2 codebase.

After thinking a bit about this, I think you are right; and we probably proceed to remove the "Exhibit B" from our codebase. Thanks for opening the issue/discussion.

> The only thing this clause prevents is compiling a GPL binary. But your version of MPLv2 still allows someone to do a worse thing: use your code in a proprietary binary while only releasing the source code of modified files. Additional proprietary code may reside in other files which are not required to be released in source code form. If you want to make sure you can get the changes back to this codebase, why not release under GPL or AGPL in the first place?
> Even if you are ok with someone building larger proprietary binary based on your source, using this variant of MPLv2 does not make sense as it only restricts someone from building GPL binaries, but not proprietary binaries.

Maybe I'm not explained this correctly; We always want preserve the MPL2 license without infecting it with GPL. **As we/you clarified previously**, this is no longer the concern and we probably go to remove this restriction from the codebase. I think all of us have the right to be wrong sometimes ;)
